drivers/spi/spi_stm32lx.c

```diff
@@ -0,0 +1,398 @@
+				  void *rx_buf, u32_t rx_buf_len)
+{
+	volatile struct spi_stm32lx *spi = SPI_STRUCT(dev);
+	struct spi_stm32lx_data *data = DEV_DATA(dev);
+	volatile u8_t *dr8 = (volatile u8_t *)&spi->dr.dr_as_u8;
+
+	__ASSERT(!((tx_buf_len && (tx_buf == NULL)) ||
+		  (rx_buf_len && (rx_buf == NULL))),
+		 "spi_stm32lx_transceive: ERROR - NULL buffer");
+
+	data->current.rx_len = rx_buf_len;
+	data->current.rx_buf = rx_buf;
+	data->current.tx_len = tx_buf_len;
+	data->current.tx_buf = tx_buf;
+
+	/* Enable Peripheral */
+	spi->cr1.bit.spe = 1;
+
+#ifdef CONFIG_SPI_STM32LX_INTERRUPT
+	if (rx_buf_len) {
+		spi->cr2.bit.rxneie = 1;
+	}
+
+	if (tx_buf_len) {
+		spi->cr2.bit.txeie = 1;
+	}
+
+	k_sem_take(&data->device_sync_sem, K_FOREVER);
+#else
+	do {
+		if (spi->sr.bit.txe && data->current.tx_len) {
+			*dr8 = *data->current.tx_buf;
+			data->current.tx_buf++;
+			data->current.tx_len--;
+		}
+
+		if (spi->sr.bit.rxne && data->current.rx_len) {
+			*data->current.rx_buf = *dr8;
+			data->current.rx_buf++;
+			data->current.rx_len--;
+		}
+
+	} while (data->current.tx_len || data->current.rx_len);
+#endif
+
+	/* Empty the RX FIFO */
+	while (spi->sr.bit.frlvl) {
+		(void)*dr8;
+	}
+
+	if (!data->current.is_slave) {
+		while (spi->sr.bit.bsy) {
+			/* NOP */
+		}
+
+		/* Disable Peripheral */
+		spi->cr1.bit.spe = 0;
+	}
+
+	return 0;
+}
+
+static const struct spi_driver_api api_funcs = {
+	.configure = spi_stm32lx_configure,
+	.slave_select = spi_stm32lx_slave_select,
+	.transceive = spi_stm32lx_transceive,
+};
+
+static inline void spi_stm32lx_get_clock(struct device *dev)
+{
+	struct spi_stm32lx_data *data = DEV_DATA(dev);
+	struct device *clk =
+		device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	__ASSERT_NO_MSG(clk);
+
+	data->clock = clk;
+}
+
+static int spi_stm32lx_init(struct device *dev)
+{
+	volatile struct spi_stm32lx *spi = SPI_STRUCT(dev);
+	struct spi_stm32lx_data *data = DEV_DATA(dev);
+	const struct spi_stm32lx_config *cfg = DEV_CFG(dev);
+
+	k_sem_init(&data->device_sync_sem, 0, UINT_MAX);
+
+	spi_stm32lx_get_clock(dev);
+
+	/* enable clock */
+	clock_control_on(data->clock,
+		(clock_control_subsys_t *)&cfg->pclken);
+
+	/* Reset config */
+	spi->cr1.val = 0;
+	spi->cr2.val = 0;
+	spi->sr.val = 0;
+
+#ifdef CONFIG_SPI_STM32LX_INTERRUPT
+	cfg->irq_config_func(dev);
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_SPI_1
+
+#ifdef CONFIG_SPI_STM32LX_INTERRUPT
+static void spi_stm32lx_irq_config_func_1(struct device *port);
+#endif
+
+static const struct spi_stm32lx_config spi_stm32lx_cfg_1 = {
+	.base = (u8_t *)SPI1_BASE,
+	.pclken = { .bus = STM32_CLOCK_BUS_APB2,
+		    .enr = LL_APB2_GRP1_PERIPH_SPI1 },
+#ifdef CONFIG_SPI_STM32LX_INTERRUPT
+	.irq_config_func = spi_stm32lx_irq_config_func_1,
+#endif
+};
+
+static struct spi_stm32lx_data spi_stm32lx_dev_data_1 = {
+};
+
+DEVICE_AND_API_INIT(spi_stm32lx_1, CONFIG_SPI_1_NAME, &spi_stm32lx_init,
+		    &spi_stm32lx_dev_data_1, &spi_stm32lx_cfg_1,
+		    PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
```

Comment:
Does this need to be initialized in PRE_KERNEL_2? Can we postpone to POST_KERNEL?